### PR TITLE
N°3136 - Add creation and modification of n-n objects in object details

### PR DESCRIPTION
**Pull request by steffunky**
Allows for the creation and modification of links in popups when accessing an object in visualization.

TODO

- [x] Move "Create" button in accordance with #222 and af804a9 principles.
- [x] In the event that CheckTo* fails, obtain the return value from the `apply_new` and `apply_modify` operations and allow retry.

### CODE REVIEW

File: application/utils.inc.php [hunk 1]
```diff
@@ -1429,30 +1429,33 @@ public static function GetPopupMenuItemsBlock(iUIBlock &$oContainerBlock, $iMenu
...
+					$aResult[] = new JSPopupMenuItem('UI:Menu:ExportPDF', Dict::S('UI:Menu:ExportPDF'), "ExportListDlg('$sOQL', '$sDataTableId', 'pdf', ".json_encode(Dict::S('UI:Menu:ExportPDF')).")");
 				}
-				$aResult[] = new JSPopupMenuItem('UI:Menu:AddToDashboard', Dict::S('UI:Menu:AddToDashboard'), "DashletCreationDlg('$sOQL', '$sContext')");
-				$aResult[] = new JSPopupMenuItem('UI:Menu:ShortcutList', Dict::S('UI:Menu:ShortcutList'), "ShortcutListDlg('$sOQL', '$sDataTableId', '$sContext')");
-				break;
+			}
+			$aResult[] = new JSPopupMenuItem('UI:Menu:AddToDashboard', Dict::S('UI:Menu:AddToDashboard'), "DashletCreationDlg('$sOQL', '$sContext')");
+			$aResult[] = new JSPopupMenuItem('UI:Menu:ShortcutList', Dict::S('UI:Menu:ShortcutList'), "ShortcutListDlg('$sOQL', '$sDataTableId', '$sContext')");
+			// TODO 3.1 Put it in its proper place
+			$aResult[] = new JSPopupMenuItem('UI:Menu:New', Dict::S('UI:Menu:New'), "LinkSetWorker.CreateLinkedObject('$sDataTableId')");
```
Comment (Molkobain) [inline application/utils.inc.php]: Shouldn't this rather be the first menu item when not in the favorite actions?

File: js/links/link_set_worker.js [hunk 2]
```diff
@@ -4,6 +4,8 @@ let LinkSetWorker = new function(){
 	const ROUTER_BASE_URL = '../pages/ajax.render.php';
```
Comment (Molkobain) [inline js/links/link_set_worker.js:4]: I know this ain't from your commit, but could you use the JS helper to make it an absolute URL? This way the script won't break if we ever move it to another folder

File: js/pages/backoffice/toolbox.js [hunk 3]
```diff
@@ -326,15 +326,17 @@ CombodoModal._InstantiateModal = function(oModalElem, oOptions) {
  */
 CombodoModal._ConvertButtonDefinition = function(aButtonsDefinitions){
 	const aConverted = [];
-	aButtonsDefinitions.forEach(element => {
-			const aButton = {
-				text: element.text,
-				class: element.class,
-				click: element.callback_on_click
+	if(aButtonsDefinitions !== null) {
```
Comment (Molkobain) [inline js/pages/backoffice/toolbox.js]: What about making it an exit condition instead? It makes it easier to read IMO

File: sources/Controller/Links/LinksetController.php [hunk 4]
```diff
@@ -99,5 +105,79 @@ public function OperationDetachLinkedObject(): \JsonPage
...
 	}
 
+	/**
+	 * @return \iTopWebPage|\AjaxPage Create edit form in its webpage
+	 * @throws \ApplicationException
+	 * @throws \ArchivedObjectException
+	 * @throws \CoreException
+	 * @throws \SecurityException
+	 */
+	public function OperationCreateLinkedObject()
```
Comment (Molkobain) [inline sources/Controller/Links/LinksetController.php:115]: Where is all that business logic copied from? We might be able to factorize part of it.

File: application/utils.inc.php (hunk 1 above)
Comment (steffunky) [reply application/utils.inc.php]: I thought it'd always be in the favorite actions as it's not an action on the list but rather a linkset action

File: js/pages/backoffice/toolbox.js (hunk 3 above)
Comment (steffunky) [reply js/pages/backoffice/toolbox.js]: Fixed in [e8c7292](https://github.com/Combodo/iTop/pull/378/commits/e8c7292e6353bb2885bec060e5d3282c96352641)

File: sources/Controller/Links/LinksetController.php (hunk 4 above)
Comment (steffunky) [reply sources/Controller/Links/LinksetController.php:115]: Copied from UILinksWidgetDirect::GetObjectCreationDlg but I'm not sure we want to mess with this bit of code

File: application/displayblock.class.inc.php [hunk 5]
```diff
@@ -2217,10 +2217,24 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, string $sI
...
+			if(isset($aExtraParams['allow_creation_in_modal']) && $aExtraParams['allow_creation_in_modal'] === true){
+				$oAddLinkActionButton = ButtonUIBlockFactory::MakeIconAction(
+					'fas fa-plus',
+					Dict::S('UI:Links:New:Button:Tooltip'),
+					Dict::S('UI:Links:New:Button:Tooltip'),
+					'',
+					false
+				);
+				$oAddLinkActionButton->AddCSSClasses(['ibo-action-button', 'ibo-regular-action-button'])
+					->SetOnClickJsCode("LinkSetWorker.CreateLinkedObject('$sId')");
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]: Discussed with Stephen IRL, we need to rework that part to avoid coupling it to linksets.

File: application/displayblock.class.inc.php [hunk 6]
```diff
@@ -2217,10 +2217,24 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, string $sI
...
 				$oActionButton = ButtonUIBlockFactory::MakeIconLink('fas fa-search', Dict::Format('UI:SearchFor_Class', MetaModel::GetName($sClass)), "{$sRootUrl}pages/UI.php?operation=search_form&do_search=0&class=$sClass{$sContext}", '', 'UI:SearchFor_Class');
 				$oActionButton->AddCSSClasses(['ibo-action-button', 'ibo-regular-action-button']);
 				$oActionsToolbar->AddSubBlock($oActionButton);
+			} 
+			
+			// - Creation in modal
+			if(isset($aExtraParams['allow_creation_in_modal']) && $aExtraParams['allow_creation_in_modal'] === true){
+				$oAddLinkActionButton = ButtonUIBlockFactory::MakeIconAction(
+					'fas fa-plus',
+					Dict::S('UI:Links:New:Button:Tooltip'),
```
Comment (Molkobain) [inline application/displayblock.class.inc.php:2231]: Discussed with Stephen IRL, will change to a label

File: css/backoffice/components/_modal.scss [hunk 7]
```diff
@@ -15,4 +24,33 @@ $ibo-modal-option--do-not-show-again--margin-top: $ibo-spacing-500 !default;
...
+  
+  &::before {
+    display: block;
+    position: absolute;
+    top: 0;
+    left: 0;
+    content: "";
+    width: $ibo-modal--is-informative--highlight--width;
+    height: $ibo-modal--is-informative--highlight--height;
+  }
```
Comment (Molkobain) [inline css/backoffice/components/_modal.scss]: Discussed with Stephen IRL, try to factorize this with the one from the alert block. This way, on next UI rework it will all be done at once.

File: sources/Application/UI/Links/AbstractBlockLinksViewTable.php [hunk 8]
```diff
@@ -62,7 +65,8 @@ public function __construct(\WebPage $oPage, \DBObject $oDbObject, string $sObje
 		$this->sAttCode = $sAttCode;
 		$this->sObjectClass = $sObjectClass;
 		$this->oDbObject = $oDbObject;
-
+		$this->sTableId = 'rel_'.$this->sAttCode;
+		$this->SetDataAttributes(['role' => 'ibo-block-links-table', 'link-attcode' => $sAttCode, 'link-class' => $this->oAttDef->GetLinkedClass()]);
```
Comment (Molkobain) [inline sources/Application/UI/Links/AbstractBlockLinksViewTable.php:69]: See if we must align `ibo-block-links-table` with the real block code (`ibo-abstract-block-links-view-table`).
Mind to update call to parent constructor as well.

File: css/backoffice/components/_modal.scss (hunk 7 above)
Comment (steffunky) [reply css/backoffice/components/_modal.scss]: Fixed in 400b155

File: sources/Application/UI/Links/AbstractBlockLinksViewTable.php (hunk 8 above)
Comment (steffunky) [reply sources/Application/UI/Links/AbstractBlockLinksViewTable.php:69]: Discussed with Guillaume IRL, it's used to have a common data-role with its child classes and a common value with block classes

File: application/displayblock.class.inc.php (hunk 6 above)
Comment (steffunky) [reply application/displayblock.class.inc.php:2231]: Fixed in 112d763

File: application/displayblock.class.inc.php (hunk 5 above)
Comment (steffunky) [reply application/displayblock.class.inc.php]: Fixed in d42f984

File: sources/Application/UI/Base/Component/DataTable/DataTable.php [hunk 9]
```diff
@@ -56,6 +56,8 @@ class DataTable extends UIContentBlock
 	 * array of data to display the first page
 	 */
 	protected $aInitDisplayData;
+	
+	protected string $sModalCreationHandler;
```
Comment (Molkobain) [inline sources/Application/UI/Base/Component/DataTable/DataTable.php:60]: Add a description in the PHPDoc of what we expect (a JS snippet) and how it will be used.

File: application/displayblock.class.inc.php [hunk 10]
```diff
@@ -2224,16 +2225,22 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, string $sI
...
 					'fas fa-plus',
 					Dict::S('UI:Links:New:Button:Tooltip'),
 					'UI:Links:New',
 					'',
 					false
 				);
+				
+				// - If we are used in a Datatable, 'datatable_' will be prefixed to our $sId so we do the same here
+				$sRealId = $sId;
+				if($this->m_sStyle === 'list' || $this->m_sStyle === 'links'|| $this->m_sStyle === 'listInObject'){
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]: Discussed IRL with Stephen, will be changed to a `in_array` instead of three conditions

File: application/displayblock.class.inc.php [hunk 11]
```diff
@@ -1802,7 +1803,7 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, string $sI
 			// Any style actions
 			// - Bulk actions on objects set
 			if ($iSetCount > 1) {
-				if ($bIsCreationAllowed && !(isset($aExtraParams['allow_creation_in_modal']) && $aExtraParams['allow_creation_in_modal'] === true)) {
+				if ($bIsCreationAllowed && !$bCreationInModalAllowed) {
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]: Discussed IRL with Stephen, will be renamed `$bIsCreationInModalAllowed` to be consistent with the other variable?

File: application/displayblock.class.inc.php [hunk 12]
```diff
@@ -2224,16 +2225,22 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, string $sI
...
 				);
+				
+				// - If we are used in a Datatable, 'datatable_' will be prefixed to our $sId so we do the same here
+				$sRealId = $sId;
+				if($this->m_sStyle === 'list' || $this->m_sStyle === 'links'|| $this->m_sStyle === 'listInObject'){
+					$sRealId = 'datatable_' . $sId;
+				}
 				$oAddLinkActionButton->AddCSSClasses(['ibo-action-button', 'ibo-regular-action-button'])
-					->SetOnClickJsCode("LinkSetWorker.CreateLinkedObject('$sId')");
+					->SetOnClickJsCode("$('#$sRealId').trigger('creation_in_modal');");
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]: Discussed IRL with Stephen, will be namespaced to `open_creation_modal.object.itop`

File: sources/Application/UI/Base/Component/DataTable/DataTableUIBlockFactory.php [hunk 13]
```diff
@@ -1076,8 +1081,10 @@ public static function GetAllowedParams(): array
 			/**give definition of id for select checkbox*/
 			'row_actions',
 			/** array of blocks displayed on every row */
-			'allow_creation_in_modal',
-			/** bool to allow a creation of a new object of this type in a modal */
+			'creation_in_modal_is_allowed',
+			/** bool to allow a creation of a new object of this type in a modal */	
+			'creation_in_modal_handler',
+			/** Handler to call when trying to create a new object in modal */
```
Comment (Molkobain) [inline sources/Application/UI/Base/Component/DataTable/DataTableUIBlockFactory.php]: Should we name it `xxx_js_handler` to explicit that it's a JS handler, not a PHP thing yo be called on the server?

File: sources/Application/UI/Base/Component/DataTable/DataTable.php (hunk 9 above)
Comment (Molkobain) [inline sources/Application/UI/Base/Component/DataTable/DataTable.php]: Add a description in the PHPDoc of what we expect (a JS snippet) and how it will be used.

File: templates/base/components/datatable/layout.ready.js.twig [hunk 14]
```diff
@@ -367,6 +367,12 @@ $('#{{ oUIBlock.GetId() }}').on('refresh.datatable.itop', function (){
     oTable{{ sListIDForVarSuffix }}.ajax.reload(null, false);
 });
 
+{% if oUIBlock.GetModalCreationHandler() is not empty %}
+$('#{{ oUIBlock.GetId() }}').on('creation_in_modal.object.itop', function (){
```
Comment (Molkobain) [inline templates/base/components/datatable/layout.ready.js.twig]: Rename as well

**Review by Molkobain — CHANGES_REQUESTED** [review]

File: application/displayblock.class.inc.php (hunk 10 above)
Comment (steffunky) [reply application/displayblock.class.inc.php]: Fixed in 22a19aa

File: application/displayblock.class.inc.php (hunk 11 above)
Comment (steffunky) [reply application/displayblock.class.inc.php]: Fixed in 686b09c

File: application/displayblock.class.inc.php (hunk 12 above)
Comment (steffunky) [reply application/displayblock.class.inc.php]: Fixed in 856b549

File: sources/Application/UI/Base/Component/DataTable/DataTableUIBlockFactory.php (hunk 13 above)
Comment (steffunky) [reply sources/Application/UI/Base/Component/DataTable/DataTableUIBlockFactory.php]: Fixed in 0bf698c

File: templates/base/components/datatable/layout.ready.js.twig (hunk 14 above)
Comment (steffunky) [reply templates/base/components/datatable/layout.ready.js.twig]: Fixed in 856b549

File: sources/Application/UI/Base/Component/DataTable/DataTable.php (hunk 9 above)
Comment (steffunky) [reply sources/Application/UI/Base/Component/DataTable/DataTable.php]: Fixed in 89fa65a

**Comment by Hipska** [comment]
Sounds great, do you have a before and after screenshot please?

File: application/utils.inc.php [hunk 15]
```diff
@@ -3045,35 +3045,18 @@ final public static function EndsWith($haystack, $needle)
...
+	 * Transform a snake_case $sInput into a CamelCase string
+	 *
 	 * @param string $sInput
 	 *
-	 * @return string Camel case representation of $sInput (eg. "something_new" becomes "SomethingNew")
+	 * @return string
 	 * @since 2.7.0
 	 */
-	public static function ToCamelCase($sInput): string
+	public static function ToCamelCase($sInput)
```
Comment (Molkobain) [inline application/utils.inc.php]: Needs to be reverted

**Review by Molkobain — CHANGES_REQUESTED** [review]

File: application/utils.inc.php (hunk 15 above)
Comment (steffunky) [reply application/utils.inc.php]: Fixed in 496d862

**Comment by steffunky** [comment]
> Sounds great, do you have a before and after screenshot please?

I don't have any before, but this is the current linkset workflow on this branch

untitled

untitled2

**Review by Molkobain — APPROVED** [review]
Nice job ✌
